src/librustc_resolve/lib.rs

```diff
@@ -4795,13 +4793,31 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
 
                 err.span_suggestion_with_applicability(
                     binding.span,
-                    rename_msg,
-                    if snippet.ends_with(';') {
-                        format!("{} as {};", &snippet[..snippet.len() - 1], suggested_name)
-                    } else {
-                        format!("{} as {}", snippet, suggested_name)
+                    &rename_msg,
+                    match (&directive.subclass, snippet.ends_with(";"), snippet.as_ref()) {
+                        (ImportDirectiveSubclass::SingleImport { .. }, false, "self") =>
+                            format!("self as {}", suggested_name),
```

Comment:
imports with `self` are a simple case, as it's only possible in the form `import foo::{ self }`